fastrand.go

```diff
@@ -42,36 +49,52 @@ func init() {
 
 // Read fills b with random data. It always returns len(b), nil.
 func (r *randReader) Read(b []byte) (int, error) {
-	// Update the progress and use the result + the seed entropy to form a
-	// unique seed to hash.
-	progress := atomic.AddUint64(&r.progress, 1)
-
-	// If it seems like the progress is about to reset, increment the first
-	// entropy bytes. Without this check, the cycle time is 2^64 iterations, but
-	// with the check it is a secure 2^128 iterations.
-	if progress > 1<<63 {
-		atomic.AddUint64((*uint64)(unsafe.Pointer(&r.entropy[0])), 1)
-		atomic.StoreUint64(&r.progress, 0)
+	// Grab a unique counter from the reader, while atomically updating the
+	// counter so that concurrent callers also end up with unique values.
+	counter := atomic.AddUint64(&r.counter, 1)
+	counterExtra := atomic.LoadUint64(&r.counterExtra)
+	// Update the second 64 bits of the counter if the first 64 bits are close
+	// to wrapping around. It is possible that the second 64 bits of the counter
+	// is updates multiple times by several concurrent threads. This wastes part
+	// of the counter space (up to 2^63 items each time), however the overall
+	// space is large enough (2^128) that wasting some every reset does not make
+	// it any more likely that the caller exhaust the whole possible search
+	// space.
+	if counter > 1<<63 {
+		atomic.AddUint64(&r.counterExtra, 1)
+		atomic.StoreUint64(&r.counter, 0)
 	}
 
-	// Copy the entropy into a separately allocated slice.
-	seed := make([]byte, 40)
-	*(*uint64)(unsafe.Pointer(&seed[0])) = progress
-	copy(seed[8:], r.entropy[:])
+	// Copy the counter and entropy into a separate slice, so that the result
+	// may be used in isolation of the other threads. The counter ensures that
+	// the result is unique to this thread.
+	seed := make([]byte, 64)
+	binary.LittleEndian.PutUint64(seed[0:8], counter)
+	binary.LittleEndian.PutUint64(seed[8:16], counterExtra)
+	// Leave 16 bytes for the inner counter.
+	copy(seed[32:], r.entropy[:])
 
-	// We now have a unique seed that we can twiddle to generate entropy.
+	// Set up an inner counter, that can be incremented to produce unique
+	// entropy within this thread.
 	n := 0
+	innerCounter := uint64(0)
+	innerCounterExtra := uint64(0)
```

Comment:
prefer `var innerCounter, innerCounterExtra uint64`

The names are a bit verbose for my tastes -- I'd be fine with `c1`, `c2`, `c3`, `c4` -- but not worth fussing over.